### PR TITLE
[Fix] correct generated type declaration

### PR DESCRIPTION
**Pull request by ocavue**
Closes https://github.com/jsx-eslint/eslint-plugin-react/issues/3838

This PR fixes various issues in the generated `index.d.ts`. The full diff for `index.d.ts` can be viewed in [this link](https://github.com/ocavue/eslint-plugin-react/commit/d8be0647b10f1ae01344bc83a8254e16064280e3). The change highlight are shown below:

```diff
  rules: {
-   'react/display-name': number;
+   'react/display-name': 2;
-   'react/jsx-key': number;
+   'react/jsx-key': 2;
```

```diff
    'jsx-no-duplicate-props': import("eslint").Rule.RuleModule;
    'jsx-no-leaked-render': import("eslint").Rule.RuleModule;
-   'jsx-no-literals': {
-      meta: import("eslint").Rule.RuleMetaData;
-      create(context: any): (false & {
-      ...
-   };
+   'jsx-no-literals': import("eslint").Rule.RuleModule;
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -55,12 +55,11 @@
     "@babel/plugin-syntax-do-expressions": "^7.24.7",
     "@babel/plugin-syntax-function-bind": "^7.24.7",
     "@babel/preset-react": "^7.24.7",
-    "@types/eslint": "=7.2.10",
     "@types/estree": "0.0.52",
     "@types/node": "^4.9.5",
     "@typescript-eslint/parser": "^2.34.0 || ^3.10.1 || ^4 || ^5 || ^6.20 || ^7.14.1 || ^8.4",
     "babel-eslint": "^8 || ^9 || ^10.1.0",
-    "eslint": "^3 || ^4 || ^5 || ^6 || ^7 || ^8 || ^9.7",
```
Comment (ocavue) [inline package.json:63]: `eslint` has built-in type declaration since version 9.10, so we don't need to install `@types/eslint` anymore.

File: lib/rules/index.js [hunk 2]
```diff
@@ -3,7 +3,7 @@
 /* eslint global-require: 0 */
 
 /** @satisfies {Record<string, import('eslint').Rule.RuleModule>} */
-module.exports = {
+const rules = {
```
Comment (ocavue) [inline lib/rules/index.js:6]: `@satisfies` doesn't work for `module.exports = {` and you have to use a variable here to make it work. That's why the missing `/** @type {import('eslint').Rule.RuleModule} */` in `lib/rules/forward-ref-uses-ref.js` wasn't be found before.

File: index.js [hunk 3]
```diff
@@ -27,6 +27,9 @@ const plugins = [
   'react',
 ];
 
+const SEVERITY_ERROR = /** @type {2} */ (2);
+const SEVERITY_OFF = /** @type {0} */ (0);
```
Comment (ljharb) [inline index.js:31]: can you elaborate on why this is needed/preferred, vs a literal number?

File: lib/rules/jsx-fragments.js [hunk 4]
```diff
@@ -170,7 +170,7 @@ module.exports = {
       ImportDeclaration(node) {
         if (node.source && node.source.value === 'react') {
           node.specifiers.forEach((spec) => {
-            if ('imported' in spec && spec.imported && spec.imported.name === fragmentPragma) {
+            if ('imported' in spec && spec.imported && 'name' in spec.imported && spec.imported.name === fragmentPragma) {
```
Comment (ljharb) [inline lib/rules/jsx-fragments.js:173]:
```suggestion
            if (
              'imported' in spec
              && spec.imported
              && 'name' in spec.imported
              && spec.imported.name === fragmentPragma
            ) {
```

File: package.json [hunk 5]
```diff
@@ -55,12 +55,11 @@
...
+    "eslint": "^3 || ^4 || ^5 || ^6 || ^7 || ^8 || ^9.10",
```
Comment (ljharb) [inline package.json:62]: unfortunately this is a breaking change, and we must stick to 9.7 as the lowest threshold.

**Review by ljharb — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (ocavue) [reply package.json:62]: There are two `^9.7` in `package.json`. One is in `peerDependencies` and another is in `devDependencies`. I only changed the one in `devDependencies`. Is it going to be a breaking change in this way?

Comment (ljharb) [reply package.json:62]: It means we won't be able to test on 9.7, 9.8, and 9.9, which means it will break and we won't know about it. Peer and dev dep ranges must always be identical.

Comment (ocavue) [reply package.json:62]: Okey let me check it back.

File: index.js (hunk 3 above)
Comment (ocavue) [reply index.js:31]: In `eslint`'s type define ([link](https://github.com/eslint/eslint/blob/7dd402d33226d821a17b22c4753ce9c40fc909bd/lib/types/index.d.ts#L898)), a severity number can only be literal `0`, `1` or `2`. 

If we have `'react/display-name': number;` in `eslint-plugin-react`'s `index.d.ts`, we will get the error below:

```
Property ''react/display-name'' is incompatible with index signature.
  Type 'number' is not assignable to type 'RuleEntry<any[]> | undefined'.ts(2322) 
```

(This error was mentioned in [this comment](https://github.com/jsx-eslint/eslint-plugin-react/issues/3838#issuecomment-2386970537))

Since `RuleEntry` is `0 | 1 | 2 | "off" | "warn" | "error"` or an array. 

Once we have a literal number `'react/display-name': 2;` than TypeScript won't complent anymore.

File: lib/rules/jsx-no-literals.js [hunk 6]
```diff
@@ -65,56 +65,10 @@ const commonPropertiesSchema = {
...
- * @typedef {ElementConfig & ElementOverrides} Config
- * @typedef {Config | OverrideConfig} ResolvedConfig
- */
-
 /**
  * Normalizes the element portion of the config
- * @param {RawConfig} config
- * @returns {ElementConfig}
+ * @param {import("../../types/rules/jsx-no-literals").RawConfig} config
+ * @returns {import("../../types/rules/jsx-no-literals").ElementConfig}
```
Comment (ocavue) [inline lib/rules/jsx-no-literals.js:71]: After adding `/** @type {import('eslint').Rule.RuleModule} */` to `lib/rules/jsx-no-literals.js`, the generated `lib/rules/jsx-no-literals.d.ts` has the following content:

```ts
declare const _exports: import('eslint').Rule.RuleModule;
export = _exports;
export type RawElementConfigProperties = {
    noStrings?: boolean;
    allowedStrings?: string[];
    ignoreProps?: boolean;
    noAttributeStrings?: boolean;
};
```

TypeScript will throw error when using this .d.ts file: `An export assignment cannot be used in a module with other exported elements.ts(2309)`

Although types like `RawElementConfigProperties` are only used internally, using JSDoc `@typedef` would export them (see issue https://github.com/microsoft/TypeScript/issues/46011) and cause the error above.

To resolve the issue, I added a new `types/rules/jsx-no-literals.d.ts` file and moved all types defined with `@typedef` into it.

File: package.json (hunk 1 above)
Comment (ocavue) [reply package.json]: I reverted this change. It seems that my PR can work anyway without touching this part.

File: index.js (hunk 3 above)
Comment (ljharb) [reply index.js:31]: I guess I’m confused why TS infers number here when it’s a literal 2, but that makes sense.

Comment (ocavue) [reply index.js:31]: > why TS infers number here when it’s a literal 2

If TS infers 2 as a literal 2, the following code would break:

```
const obj = { a: 1, b: 2 }
obj.b = 3 // Cannot assign 3 to 2
```

TS provides a handy tool `as const` to address such issue, sadly we don't have similar thing in JSDoc.

```ts
const obj = { a: 1, b: 2 }
     // ^: { a: number; b: number }
const obj2 = { a: 1, b: 2 } as const 
     // ^: { a: readonly 1; b: readonly 2 }
```

Comment (nchevsky) [reply index.js:31]: > TS provides a handy tool `as const` to address such issue, sadly we don't have similar thing in JSDoc.

We actually do, since [TypeScript v4.5](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#const-assertions-and-default-type-arguments-in-jsdoc):

https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#casts

Comment (ocavue) [reply index.js:31]: Thank you for the correction! I tried that but it didn't work. I realized that's because this project is still using TS v3.  Since the current approach works I'd prefer not updating TS in this PR.

**Comment by ljharb** [comment]
ok, i fixed the issue with index.js in master, but i've rebased this PR on top of that, since this is probably a better outcome still :-)

File: types/string.prototype.repeat/index.d.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
 declare module 'string.prototype.repeat' {
-  export = typeof Function.call.bind(String.prototype.repeat);
+  function repeat(text: string, count: number): string;
```
Comment (ljharb) [inline types/string.prototype.repeat/index.d.ts:2]: this means that it won't always inherit the builtin type for repeat. is this necessary?

File: lib/rules/jsx-no-literals.js [hunk 8]
```diff
@@ -485,7 +447,7 @@ module.exports = {
               });
           });
       },
-    } : false, {
+    } : {}, {
```
Comment (ljharb) [inline lib/rules/jsx-no-literals.js:450]: why this change? primitives are object-spread the same as `{}`.

File: types/string.prototype.repeat/index.d.ts (hunk 7 above)
Comment (ocavue) [reply types/string.prototype.repeat/index.d.ts:2]: The previous version of `types/string.prototype.repeat/index.d.ts` doesn't work. 

```ts
declare module 'string.prototype.repeat' {
  export = typeof Function.call.bind(String.prototype.repeat);
           // ^: The expression of an export assignment must be an identifier or qualified name in an ambient context.ts(2714)
}
```

This error didn't show up because `tsconfig.json` didn't include `types/` at all, thus `string.prototype.repeat` was been treated as `any` by TypeScript.

I'm not sure if there are better way to write this `index.d.ts`, but it definitely need some changes.

Comment (ljharb) [reply types/string.prototype.repeat/index.d.ts:2]: ah, right. how about:
```suggestion
  const repeat: typeof Function.call.bind(String.prototype.repeat);
```
?

File: lib/rules/jsx-no-literals.js (hunk 8 above)
Comment (ocavue) [reply lib/rules/jsx-no-literals.js:450]: I don't remember why I did this change exactly, probably because `false` isn't compatible with the latest TypeScript or `@types/eslint`. 

Anyway, let me revert this change for now. We might will need to change it to `{}` in the future if we update TypeScript or `@types/eslint`.

**Comment by ocavue** [comment]
@ljharb Hi. Thanks for your review. I've addressed all your previous comments.

In addition to that, I was thinking why we have issues like https://github.com/jsx-eslint/eslint-plugin-react/issues/3838 even if we have `.github/workflows/type-check.yml` to check the built types. I found there are three issues in the CI workflow and I fixed them. 

1. In `test-published-types/index.js`, to make the typescript to check the type, we need to create a variable first and export it later

```ts
// Bad
/** @type {import('eslint').Linter.Config[]} */
module.exports = [ ... ] 

// Good
/** @type {import('eslint').Linter.Config[]} */
const config = [ ... ]
module.exports = config;
```

2. In `test-published-types/package.json`, if we use `"eslint-plugin-react": "file:.."`, we would copy/link all node_modules under eslint-plugin-react to `test-published-types/node_modules/eslint-plugin-react/node_modules`. This is problematic because `test-published-types` use the built-in types from the latest `eslint`, while `eslint-plugin-react` use the deprecated `@types/eslint`. They have conflict. 
  
   I fixed this by packing `eslint-plugin-react` into a `.taz` file and install it later. By doing that, `@types/eslint`, which is a dev dependency of `eslint-plugin-react`, won't be installed under `test-published-types`.

3. There is a small issue in the eslint's type declarations:
   
```bash
$ cd test-published-types
$ npx tsc --lib es2015
   
node_modules/eslint/lib/types/index.d.ts:928:81 - error TS2574: A rest element type must be an array type.

928     type RuleSeverityAndOptions<Options extends any[] = any[]> = [RuleSeverity, ...Partial<Options>];
                                                                                    ~~~~~~~~~~~~~~~~~~~
```

I need to fix it by adding `"skipLibCheck": false` in `test-published-types/tsconfig.json`.

File: types/string.prototype.repeat/index.d.ts (hunk 7 above)
Comment (ocavue) [reply types/string.prototype.repeat/index.d.ts:2]: It doesn't work :(

```
declare module 'string.prototype.repeat' {
  const repeat: typeof Function.call.bind(String.prototype.repeat);
                                      // ^: ',' expected. ts(1005)
  export = repeat;
}
```

**Comment by burtek** [comment]
Any updates on this? :eyes:

File: lib/rules/forbid-foreign-prop-types.js [hunk 9]
```diff
@@ -109,7 +109,7 @@ module.exports = {
             && !ast.isAssignmentLHS(node)
             && !isAllowedAssignment(node)
           )) || (
-            // @ts-expect-error The JSXText type is not present in the estree type definitions
+            // @ts-expect-error: Literal is not a valid type
```
Comment (ljharb) [inline lib/rules/forbid-foreign-prop-types.js:112]: O.o why is Literal not a valid type?

File: types/string.prototype.repeat/index.d.ts (hunk 7 above)
Comment (ljharb) [reply types/string.prototype.repeat/index.d.ts:2]: i can fix this later, probably by making a DT package.

**Review by ljharb — APPROVED** [review]

File: lib/rules/forbid-foreign-prop-types.js (hunk 9 above)
Comment (ocavue) [reply lib/rules/forbid-foreign-prop-types.js:112]: Sorry. It's `JSXText` that's not a valid type.

```
(node.property.type === 'Literal' || node.property.type === 'JSXText')
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                     ^: This comparison appears to be unintentional because the 
                                        types '"ArrayExpression" | "ArrowFunctionExpression" | 
                                        "AssignmentExpression" | "AwaitExpression" | 
                                        "BinaryExpression" | "CallExpression" | "ChainExpression" | 
                                        "ClassExpression" | ... 16 more ... | "PrivateIdentifier"' 
                                        and '"JSXText"' have no overlap. 
                                        ts(2367)
```

Comment (ocavue) [reply lib/rules/forbid-foreign-prop-types.js]: I have reverted this comment change.
